adding virtualhostoption docs (#8941)
solo-io/gloo: docs/content/guides/traffic_management/request_processing/delegation/_index.md

---
title: Delegate VirtualHost and RouteTable options
weight: 200
description: Define 
---

You can use the `VirtualHostOption` and `RouteTableOption` resources to specify option settings for virtual hosts and route tables, such as traffic management configurations or policies. The main virtual host and route table resources can then delegate to these options to apply them. 

> This approach is useful if you have traffic management or policy settings that you want to apply to multiple virtual hosts or routes, such as a rate limiting configuration that you want to reuse across hosts. It can also help to keep the main virtual host and route table resources small which can simplify the troubleshooting process in case of an error.

Reviewer: ```suggestion
This approach is useful if you have traffic management or policy settings that you want to apply to multiple virtual hosts or routes, such as a rate limiting configuration that you want to reuse across hosts. It can also help to keep the main virtual host and route table resources concise, which can simplify the troubleshooting process in case of an error.
```

## Rules for delegating to VirtualHostOption and RouteTableOption resources {#rules}

When delegating to `VirtualHostOption` and `RouteTableOption` resources from the main virtual service or route table resource, the following rules apply: 
- Options that are specified on the main virtual service or route table resource take precedence over options that are specified on `VirtualHostOption` and `RouteTableOption` resources. For example, if a header manipulation policy is specified on the main virtual service resource, header manipulation policies in `VirtualHostOption` resources that the virtual service delegates to are ignored. 
- If delegating to multiple `VirtualHostOption` and `RouteTableOption` resources, the resource configuration is applied in order. For example, if a virtual service delegates to two `VirtualHostOption` resources, and both of them specify retry policies, only the retry policy from the `VirtualHostOption` resource that is delegated to first is applied. The retry policy in subsequent `VirtualHostOption` resources is ignored. 

{{% notice note %}}

> Keep in mind that delegating to a lot of `VirtualHostOption` and `RouteTableOption` resources can complicate the troubleshooting process as you might need to analyze multiple option resources to find the root cause of an issue. In addition, updates to `VirtualHostOption` and `RouteTableOption` resources that introduce errors can have large impacts on your environment depending on how often these resources are delegated to. 

Reviewer: ```suggestion
Keep in mind that delegating to several `VirtualHostOption` and `RouteTableOption` resources can complicate the troubleshooting process, as you might need to analyze multiple option resources to find the root cause of an issue. In addition, updates to `VirtualHostOption` and `RouteTableOption` resources that introduce errors can have large impacts on your environment depending on how often these resources are delegated to. 
```

     namespace: gloo-system
   spec:
     options:
       headerManipulation:
         requestHeadersToRemove: ["content-header"]
       ratelimitBasic:
         anonymousLimits:
           unit: MINUTE
           requestsPerUnit: 1
       retries:
         retryOn: 'connect-failure'
         numRetries: 5
         perTryTimeout: '10s'
   EOF
   ```

6. Verify your final settings by sending multiple requests to the httpbin app. Following the [rules for applying virtual host and route table option resources](#rules), the following behavior is expected: 
   - `headerManipulation`: Because the main virtual service resource specifies header manipulation settings, these settings take precedence over header manipulation settings that are configured on delegated `VirtualHostOption` resources. Consequently, the header manipulation settings from both `VirtualHostOption` resources are ignored. 
   - `retries`: Retry settings from the `virtualhost-options-1` resource take precedence over the retry settings in `virtualhost-options-2`, because the main virtual service resource delegates to `virtualhost-options-1` first. Retry settings in subsequent `VirtualHostOption` resources are ignored. 

>    - `ratelimitBasic`: Because the main virtual service and the `virtualhost-options-1` virtual host option resources do not specify any rate limiting settinge, the rate limit settings from `virtualhost-options-2` are applied to the virtual host. 

Reviewer: ```suggestion
   - `ratelimitBasic`: Because the main virtual service and the `virtualhost-options-1` virtual host option resources do not specify any rate limiting settings, the rate limit settings from `virtualhost-options-2` are applied to the virtual host. 
```
